Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -1429,6 +1468,114 @@ TEST_F(SerializerTest, EncodesResumeTokens) {
   ExpectRoundTrip(model, proto);
 }
 
+TEST_F(SerializerTest, EncodesListenRequestLabels) {
+  core::Query q = Query("docs");
+
+  std::map<QueryPurpose, std::unordered_map<std::string, std::string>>
+      purpose_to_label = {
+          {QueryPurpose::Listen, {}},
+          {QueryPurpose::LimboResolution,
+           {{"goog-listen-tags", "limbo-document"}}},
+          {QueryPurpose::ExistenceFilterMismatch,
+           {{"goog-listen-tags", "existence-filter-mismatch"}}},
+      };
+
+  for (const auto& p : purpose_to_label) {
+    QueryData model(q, 1, 0, p.first);
+
+    auto result = serializer.EncodeListenRequestLabels(model);
+    EXPECT_EQ(result, p.second);
+  }
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithTargetChange) {
+  WatchTargetChange model(WatchTargetChangeState::Added, {1, 2},
+                          ByteString("resume_token"));
+
+  v1::ListenResponse proto;
+
+  proto.mutable_target_change()->set_target_change_type(
+      v1::TargetChange_TargetChangeType::TargetChange_TargetChangeType_ADD);
+  proto.mutable_target_change()->add_target_ids(1);
+  proto.mutable_target_change()->add_target_ids(2);
+  proto.mutable_target_change()->set_resume_token("resume_token");
+
+  SCOPED_TRACE("DecodesListenResponseWithTargetChange");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithDocumentChange) {
+  SnapshotVersion version = Version(123456789L);
+  DocumentWatchChange model(
+      {1, 3}, {2, 4}, Key("one/two/three/four"),
+      Doc("one/two/three/four", 123456789L, Map("foo", "bar")));
+
+  v1::ListenResponse proto;
+
+  auto document_change = proto.mutable_document_change();
+  document_change->mutable_document()->set_name(
+      "projects/p/databases/d/documents/one/two/three/four");
+  document_change->mutable_document()->mutable_update_time()->set_seconds(
+      version.timestamp().seconds());
+  document_change->mutable_document()->mutable_update_time()->set_nanos(
+      version.timestamp().nanoseconds());
+  (*document_change->mutable_document()->mutable_fields())["foo"] =
+      ValueProto("bar");
+
+  document_change->add_target_ids(1);
+  document_change->add_target_ids(3);
+  document_change->add_removed_target_ids(2);
+  document_change->add_removed_target_ids(4);
+
+  SCOPED_TRACE("DecodesListenResponseWithDocumentChange");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithDocumentDelete) {
```

Comment:
Similarly, there are [4 tests](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Remote/FSTSerializerBetaTests.mm#L910-L966) involving `DocumentWatchChange` in Objective-C, can you please port them all?

In general, it's okay to add extra tests, but I think all existing tests should be ported.